jerry-core/jerry.c

```diff
@@ -651,893 +702,885 @@ jerry_create_error (jerry_error_t error_type, /**< type of error */
+{
+  prop_desc_p->is_value_defined = false;
+  prop_desc_p->value = ecma_make_simple_value (ECMA_SIMPLE_VALUE_UNDEFINED);
+  prop_desc_p->is_writable_defined = false;
+  prop_desc_p->is_writable = false;
+  prop_desc_p->is_enumerable_defined = false;
+  prop_desc_p->is_enumerable = false;
+  prop_desc_p->is_configurable_defined = false;
+  prop_desc_p->is_configurable = false;
+  prop_desc_p->is_get_defined = false;
+  prop_desc_p->getter = ecma_make_simple_value (ECMA_SIMPLE_VALUE_UNDEFINED);
+  prop_desc_p->is_set_defined = false;
+  prop_desc_p->setter = ecma_make_simple_value (ECMA_SIMPLE_VALUE_UNDEFINED);
+} /* jerry_init_property_descriptor */
+
+/**
+ * Define a property to the specified object with the given name.
  *
  * Note:
- *      If current code is executed on top of interpreter, using is_direct argument,
- *      caller can enable direct eval mode that is equivalent to calling eval from
- *      within of current JS execution context.
+ *      returned value must be freed with jerry_release_value, when it is no longer needed.
  *
- * @return completion status
+ * @return true  - if success
+ *         thrown error - otherwise
  */
-jerry_completion_code_t
-jerry_eval (const jerry_char_t *source_p, /**< source code */
-            size_t source_size, /**< length of source code */
-            bool is_direct, /**< perform eval invocation in direct mode */
-            bool is_strict, /**< perform eval as it is called from strict mode code */
-            jerry_value_t *retval_p) /**< [out] returned value */
+jerry_value_t
+jerry_define_own_property (const jerry_value_t obj_val, /**< object value */
+                           const jerry_value_t prop_name_val, /**< property name (string value) */
+                           const jerry_property_descriptor_t *prop_desc_p) /**< property descriptor */
 {
   jerry_assert_api_available ();
 
-  jerry_completion_code_t status;
+  if (ECMA_IS_VALUE_ERROR (obj_val)
+      || ECMA_IS_VALUE_ERROR (prop_name_val))
+  {
+    return ecma_raise_type_error (ECMA_ERR_MSG (error_value_msg_p));
+  }
+  else if (!ecma_is_value_object (obj_val)
+           || !ecma_is_value_string (prop_name_val))
+  {
+    return ecma_raise_type_error (ECMA_ERR_MSG (wrong_args_msg_p));
+  }
 
-  ecma_value_t completion = ecma_op_eval_chars_buffer ((const lit_utf8_byte_t *) source_p,
-                                                       source_size,
-                                                       is_direct,
-                                                       is_strict);
+  ecma_property_descriptor_t prop_desc = ecma_make_empty_property_descriptor ();
 
-  status = jerry_convert_eval_completion_to_retval (completion, retval_p);
+  prop_desc.is_value_defined = prop_desc_p->is_value_defined;
+  prop_desc.value = prop_desc_p->value;
 
-  return status;
-} /* jerry_eval */
+  prop_desc.is_writable_defined = prop_desc_p->is_writable_defined;
+  prop_desc.is_writable = prop_desc_p->is_writable;
```

Comment:
You cannot trust the is_writable flag. The correct form is:

    prop_desc.is_writable_defined = prop_desc_p->is_writable_defined;
    prop_desc.is_writable = prop_desc_p->is_writable_defined ? prop_desc_p->is_writable : false;

Similar cases below.